core/src/mfd.rs

```diff
@@ -0,0 +1,74 @@
+//! Implementation of a memory fd
+
+use std::io::{Read, Seek, SeekFrom, Write};
+use std::os::fd::{AsRawFd, RawFd};
+
+use anyhow::{bail, Result};
+use memfd::{FileSeal, Memfd, MemfdOptions};
+
+pub struct Mfd(Memfd);
+
+impl Mfd {
+    /// Creates an empty named Memfd
+    pub fn create(name: &str) -> Result<Self> {
+        let opts = MemfdOptions::default().allow_sealing(true);
+        let mfd = opts.create(name)?;
+        Ok(Self(mfd))
+    }
+
+    /// Reads all data available
+    pub fn read_all(&self) -> Result<Vec<u8>> {
+        self.0.as_file().seek(SeekFrom::Start(0))?;
+        let mut buf: Vec<u8> = Vec::with_capacity(128);
+        self.0.as_file().read_to_end(&mut buf)?;
+
+        Ok(buf)
+    }
+
+    /// Wipes the mfd and overwrites it
+    pub fn write(&mut self, data: &[u8]) -> Result<()> {
+        self.0.as_file().seek(SeekFrom::Start(0))?;
+        self.0.as_file().set_len(0)?;
+        self.0.as_file().write_all(data)?;
+        Ok(())
+    }
+
+    /// Finalizes the mfd so that it becomes immutable
+    pub fn finalize(&mut self) -> Result<()> {
+        self.0.add_seal(FileSeal::SealShrink)?;
+        self.0.add_seal(FileSeal::SealGrow)?;
+        self.0.add_seal(FileSeal::SealWrite)?;
+        self.0.add_seal(FileSeal::SealSeal)?;
+        Ok(())
+    }
```

Comment:
You have a suggest for an alternative interface? I personally do not consider myself able to design one with that scope.

Reply:
IMHO there is two cases:

the memfd from partition to hypervisor should be `SEAL_SHRINK|SEAL_GROW|SEAL_WRITE`. This ensures that the hypervisor can happily read the memfd, in multiple passes, but it will never change size or data.

The second memfd from hypervisor to partition should be `SEAL_SHRINK|SEAL_GROW|SEAL_SEAL`. This ensures the hypervisor can write the response, and no one can take away write access from it.

Reply:
How about a bitmask as it's parameter? Or an enum for each of the cases.

Reply:
i would probably prefer an enum with the cases: 
- Unsealed
- SealedReadable
- SealedWritable

Care needs to be taken with the `from_fd` function as it needs to check the seales on the memfd for providing the correct inner enum (or error in case the seals do not match any of these).

Reply:
I've done something in the top-most commit. I personally do not like this solution, because I personally see no usecase other than `Seals::Readable`.